// @yandex/ymaps3-web-mercator-projection/dist/esm/web-mercator.js
import { cycleRestrict, restrict, DEG_TO_RAD, RAD_TO_DEG } from './utils';
/**
 * @class Mercator projection onto a sphere.
 * Used by many mapping services, in particular OpenStreetMap.
 * @name projection.webMercator
 * @augments Projection
 * @static
 * @example
 * ```js
 * const {YMap, YMapTileDataSource, YMapLayer} = ymaps3;
 * const {WebMercator} = await ymaps3.import('@yandex/ymaps3-web-mercator-projection');
 *
 * // Create a map in the web Mercator projection
 * const map = new YMap(document.getElementById('map_id'), {
 *     // ...
 *     projection: new WebMercator()
 * });
 *
 * // Add osm tiles
 * const dataSource = new YMapTileDataSource({
 *     id: 'osmSource',
 *     zoomRange: {min: 0, max: 19},
 *     clampMapZoom: true,
 *     raster: {
 *         type: 'ground',
 *         fetchTile: 'https://tile.openstreetmap.org/{{z}}/{{x}}/{{y}}.png'
 *     },
 *     copyrights: ['© OpenStreetMap contributors']
 * });
 *
 * const layer = new YMapLayer({
 *     id: 'osm',
 *     source: 'osmSource',
 *     type: 'ground',
 * });
 * ```
 */
export class WebMercator {
    constructor() {
        this._maxLatitudeRad = Math.atan(Math.sinh(Math.PI));
        this.type = 'EPSG:3857';
    }
    toWorldCoordinates(coords) {
        return {
            x: this._longitudeToWorldX(coords[0]),
            y: this._latitudeToWorldY(coords[1])
        };
    }
    fromWorldCoordinates(world) {
        return [this._worldXToLongitude(world.x), this._worldYToLatitude(world.y)];
    }
    // For all transformations, we don't need either the radius or the equator, only their ratio, which is always Math.PI
    _worldXToLongitude(x) {
        return cycleRestrict(x * Math.PI * RAD_TO_DEG, -180, 180);
    }
    _longitudeToWorldX(lng) {
        return cycleRestrict((lng * DEG_TO_RAD) / Math.PI, -1, 1);
    }
    _latitudeToWorldY(lat) {
        // We restrict latitude to use it with an aspect ratio of one and prevent Infinity calculations
        const latitude = restrict(lat * DEG_TO_RAD, -this._maxLatitudeRad, this._maxLatitudeRad);
        const tanTemp = Math.tan(Math.PI / 4 + latitude / 2);
        return restrict(Math.log(tanTemp) / Math.PI, -1, 1);
    }
    _worldYToLatitude(y) {
        const merkatorYDivR = -y * Math.PI;
        const latitude = Math.PI / 2 - 2 * Math.atan(Math.exp(merkatorYDivR));
        return restrict(latitude * RAD_TO_DEG, -90, 90);
    }
}
